### ISSUE TITLE
Unexpectedly ignored words in @import

### ISSUE THREAD

**Issue by lydell** [shellscape/postcss-less#88]
* Node Version: 8.3.0
* NPM Version: 5.3.0
* postcss-less Version: 1.1.0

### LESS

```less
@import missing "missing" "not missing";
```

### JavaScript

```js
require("postcss-less").parse('@import missing "missing" "not missing";')
```

### Expected Behavior

Error thrown, or `missing` and `"missing"` being somewhere on the `Import` node.

Note: The example code is invalid Less ("malformed import statement").

### Actual Behavior

The example Less code parses the same as this:

```less
@import "not missing";
```

In other words, `missing` and `"missing"` are totally ignored.

### How can we reproduce the behavior?

```js
require("postcss-less").parse('@import missing "missing" "not missing";')
```

### Background

This causes problems for [Prettier](https://github.com/prettier/prettier/) when people forget the semicolon after the `@import`: prettier/prettier#2584

```less
@import "../assets/font/league-gothic/league-gothic.css"

.ManagerPage {
  height: 100%;
}
```

... is turned into this by Prettier:

```less
@import .ManagerPage {
  height: 100%;
}
```

... because `"../assets/font/league-gothic/league-gothic.css"` is entirely missing from the AST. Prettier would need:

- Either an error to be thrown
- or `"../assets/font/league-gothic/league-gothic.css"` being somewhere on the `Import` node so that it can be printed back.

**Comment by shellscape** [comment]
hm that's a bummer. in these kinds of situations it's helpful to have an example of the AST if they AST isn't correct, or is missing things. perfectly acceptable and encouraged to provide that via gist or inline if it's not too bananas.

so we need a few questions answered here:
1. what's the AST look like in the broken example?
2. [the spec](https://drafts.csswg.org/css-cascade-3/#at-ruledef-import) suggests that the semicolon is required, but afaik LESS doesn't require it. is that the case or does LESS actually require a semicolon?
3. the spec suggests that anything goes after the `@import` but before the `;`. So how do we represent that in an AST?
4. does postcss proper handle importers in a better way, and can we piggy back on that?

**Comment by lydell** [comment]
1.

Here's the AST from `postcss-less`:

```
Root {
  raws: { semicolon: false, after: '' },
  type: 'root',
  nodes:
   [ Import {
       raws: { before: '', afterName: ' ', after: ' ', semicolon: true },
       type: 'import',
       nodes: [],
       name: 'import',
       parent: [Circular],
       source:
        { start: { line: 1, column: 1 },
          input:
           Input {
             css: '@import missing "missing" "not missing";',
             id: '<input css 1>' },
          end: { line: 1, column: 40 } },
       importPath: '"not missing"',
       directives: '' } ],
  source:
   { input:
      Input {
        css: '@import missing "missing" "not missing";',
        id: '<input css 1>' },
     start: { line: 1, column: 1 } } }
```

---

2.

AFAIK, the semicolon is _required_ in CSS, SCSS and Less. But none of the postcss parsers throw errors for missing semicolons. But the semicolon is a bit of a red herring for this issue.

Also, both `postcss` and `postcss-scss` parse like the following as a single at-rule with sub-nodes (just like `postcss-less`):

```css
@import "../assets/font/league-gothic/league-gothic.css"

.ManagerPage {
  height: 100%;
}
```

---

3.

> So how do we represent that in an AST?

That's what we're here to discuss :)

---

4.

Both `postcss` and `postcss-scss` parse like this:

```
Root {
  raws: { semicolon: true, after: '' },
  type: 'root',
  nodes:
   [ AtRule {
       raws: { before: '', between: '', afterName: ' ' },
       type: 'atrule',
       name: 'import',
       parent: [Circular],
       source:
        { start: { line: 1, column: 1 },
          input:
           Input {
             css: '@import missing "missing" "not missing";',
             id: '<input css 5>' },
          end: { line: 1, column: 40 } },
       params: 'missing "missing" "not missing"' } ],
  source:
   { input:
      Input {
        css: '@import missing "missing" "not missing";',
        id: '<input css 5>' },
     start: { line: 1, column: 1 } } }
```

In other words, they don't do anything special for `@import`s – it's parsed as a general at-rule.

**Comment by shellscape** [comment]
Interesting. I could see this being a valuable addition to postcss once complete.

I'm open to suggestions as to how to represent the other bits that are being dropped from the Node.  We should follow the spec as closely as possible, while making adjustments for LESS (with things such as [directives](http://lesscss.org/features/#import-directives-feature), which don't exist in the spec)

```css
@import [ <url> | <string> ] <media-query-list>? ;
```

**Comment by lydell** [comment]
> I could see this being a valuable addition to postcss once complete.

What does "this" mean here? :)

**Comment by lydell** [comment]
Copying in examples from the spec for convenience:

```css
@import url("fineprint.css") print;
@import url("bluish.css") projection, tv;
@import url("narrow.css") handheld and (max-width: 400px);
```

**Comment by shellscape** [comment]
"this" would be a fully functioning `ImportNode` that is handled differently than `AtRule`.

**Comment by lydell** [comment]
I not sure that makes sense. As far as I know, postcss does not treat any at-rule in specially, and just stores everything after `@foo` as one long string. [postcss-media-query-parser](https://github.com/dryoma/postcss-media-query-parser) already seems to parse `url("narrow.css") handheld and (max-width: 400px)` really well.

**Comment by shellscape** [comment]
Storing `<media-query-list>?` in a property for further inspection (by something like postcss-media-query-parser) makes a lot of sense to me.

**Comment by shellscape** [comment]
@lydell ping. any interest in a PR for this one?

**Comment by lydell** [comment]
No, no interest. I wish I had, though.
